File: haskell/rna-transcription/dna.hs
module DNA (toRNA) where

toRNA :: [Char] -> [Char]
toRNA x = map subNucleotide x
  where subNucleotide n = case (n) of 
          ('G') -> 'C'
          ('C') -> 'G'
          ('T') -> 'A'
          ('A') -> 'U'
          _     -> 'X'

File: scala/rna-transcription/src/main/scala/Dna.scala
object Dna {
  val rna = Map(
    'G' -> 'C',
    'C' -> 'G',
    'T' -> 'A',
    'A' -> 'U'
  )

  def toRna(sequence: String): Option[String] = {
    val getRna: (Option[String], Char) => Option[String] = {
      case (Some(translated), c) if rna.keySet.contains(c) =>
        Some(translated + rna(c))
      case (_, _) =>
        None
    }
    sequence.foldLeft(Option(""))(getRna)
  }
}
